### PR TITLE
54 amb bridge first pass

### PR DESCRIPTION
**Pull request by CJBangert**
## What

- added details about how the Arbitrary message bridge works
 - added detailed diagram depicting how an arbitrary function call happens cross-chain
 - kicked out majority of details about mediator contracts to Omnibridge section, working on that in separate branch 
 - re-organized sidebar to have AMB bridge section first, then xdai bridge and omnibridge as they build off the AMB Bridge 

## Refs

- #54

### CODE REVIEW

**Comment by alebanzas** [comment]
@dan-gnosis could you review it? thanks!!

File: docs/bridges/tokenbridge/amb-bridge.md [hunk 1]
```diff
@@ -19,28 +19,23 @@ The AMB currently supports Ethereum and Binance Smart Chain, and is part of the
...
 
 ### Overview
+|                       | Detail                                                                |
+|-----------------------|-----------------------------------------------------------------------|
+| Frontend URL          | N/A                                                                   |
+| Trust Model           | [4-of-6 Validator Multisig](#bridge-validators)                       |
+| Governance            | [7-of-16 Multisig](#bridge-governance)                                |
+| Governance Parameters |                                                                       |
+| Bug Bounty            | [up to $2m](https://immunefi.com/bounty/gnosischain/)                 |
+| Bug Reporting         | On the Immunefi site [here](https://immunefi.com/bounty/gnosischain/) |
```
Comment (dan-gnosis) [inline docs/bridges/tokenbridge/amb-bridge.md:29]: [Immunefi](https://immunefi.com/bounty/gnosischain/)

File: docs/bridges/tokenbridge/amb-bridge.md [hunk 2]
```diff
@@ -57,16 +52,104 @@ References:
...
 
-### Bridge Revenue
+## Key Contracts
 
-### Analytics
+### Ethereum
+
+| Contract                            | Address                                    |
+|-------------------------------------|--------------------------------------------|
+| AMB/Omnibridge Multi-Token Mediator | 0x88ad09518695c6c3712AC10a214bE5109a655671 |
```
Comment (dan-gnosis) [inline docs/bridges/tokenbridge/amb-bridge.md:61]: Link Etherscan address (preferably the page where you can interact with the contract)

File: docs/bridges/tokenbridge/amb-bridge.md [hunk 3]
```diff
@@ -57,16 +52,104 @@ References:
...
+| AMB Contract Proxy (Foreign)        | 0x4C36d2919e407f0Cc2Ee3c993ccF8ac26d9CE64e |
+
+### Gnosis
+| Contract                            | Address                                    |
+|-------------------------------------|--------------------------------------------|
+| AMB/Omnibridge Multi-Token Mediator | 0xf6A78083ca3e2a662D6dd1703c939c8aCE2e268d |
```
Comment (dan-gnosis) [inline docs/bridges/tokenbridge/amb-bridge.md:67]: Link Gnosisscan address, e.g. 

https://gnosisscan.io/address/0xf6a78083ca3e2a662d6dd1703c939c8ace2e268d#code

File: docs/bridges/tokenbridge/amb-bridge.md [hunk 4]
```diff
@@ -57,16 +52,104 @@ References:
...
+| AMB Contract Proxy (Home)           | 0x75Df5AF045d91108662D8080fD1FEFAd6aA0bb59 |
 
-References: 
 ## How it works
+### Terminology
+* __Home (Native) Network__: Of the two networks being bridged between, the home or native network is the one with fast and inexpensive operations. All bridge operations to collect validator confirmations are performed on this side of the bridge.
```
Comment (dan-gnosis) [inline docs/bridges/tokenbridge/amb-bridge.md:72]: This is actually quite concerning from a Bridge Design perspective. That means that reverse bridge-out operations are wholly witnessed on Gnosis Chain, with no signatories on Ethereum L1.

File: docs/bridges/tokenbridge/amb-bridge.md [hunk 5]
```diff
@@ -57,16 +52,104 @@ References:
...
+                                bytes _data,
+                                uint256 _gas) external;
+```
+ | param      | details                                                                                                   |
+ |------------|-----------------------------------------------------------------------------------------------------------|
+ | \_contract | address of contract on other network                                                                      |
+ | \_data     | encoded bytes of the method selector and the params that will be called in the contract on the other side |
+ | \_gas      | gas to be provided in execution of the method call on the other side                                      |
+ 
+ ![](/img/bridges/diagrams/amb-bridge-contract-flow.png)
```
Comment (dan-gnosis) [inline docs/bridges/tokenbridge/amb-bridge.md:88]: Where is the contract which holds this method?

I don't see it in: 
https://gnosisscan.io/address/0xf6a78083ca3e2a662d6dd1703c939c8ace2e268d#readProxyContract

EDIT: 
I verified the contract as a proxy on Gnosisscan.io, and was able to find the method below

https://gnosisscan.io/address/0x75Df5AF045d91108662D8080fD1FEFAd6aA0bb59#writeProxyContract

File: docs/bridges/tokenbridge/amb-bridge.md [hunk 6]
```diff
@@ -57,16 +52,104 @@ References:
...
+| AMB/Omnibridge Multi-Token Mediator | 0xf6A78083ca3e2a662D6dd1703c939c8aCE2e268d |
+| AMB Contract Proxy (Home)           | 0x75Df5AF045d91108662D8080fD1FEFAd6aA0bb59 |
 
-References: 
 ## How it works
+### Terminology
+* __Home (Native) Network__: Of the two networks being bridged between, the home or native network is the one with fast and inexpensive operations. All bridge operations to collect validator confirmations are performed on this side of the bridge.
+* __Foreign Network__: Can be any EVM chain, generally it refers to Ethereum.
+
+### Call a cross-chain method via AMB:
```
Comment (dan-gnosis) [inline docs/bridges/tokenbridge/amb-bridge.md:75]: I highly recommend we also add this, later on, as a tutorial in the Developer Docs

File: docs/bridges/tokenbridge/amb-bridge.md [hunk 7]
```diff
@@ -57,16 +52,104 @@ References:
...
+ | param      | details                                                                                                   |
+ |------------|-----------------------------------------------------------------------------------------------------------|
+ | \_contract | address of contract on other network                                                                      |
+ | \_data     | encoded bytes of the method selector and the params that will be called in the contract on the other side |
+ | \_gas      | gas to be provided in execution of the method call on the other side                                      |
+ 
+ ![](/img/bridges/diagrams/amb-bridge-contract-flow.png)
+
+#### Foreign Network -> Home Network
+1. User calls `foo()` on an originator contract
```
Comment (dan-gnosis) [inline docs/bridges/tokenbridge/amb-bridge.md:91]: What is an originator contract? Needs clearer term

File: docs/bridges/tokenbridge/amb-bridge.md [hunk 8]
```diff
@@ -57,16 +52,104 @@ References:
...
+2. Originator contract calls `requireToPassMessage()` on Foreign Bridge contract, and encodes `foo()`, target address, and includes some tokens for gas.
```
Comment (dan-gnosis) [inline docs/bridges/tokenbridge/amb-bridge.md:92]: Does the "Foreign Bridge contract" here use the same terminology as above (e.g. Home, vs Foreign)? 

Longer term, I think we need to clean up some of the nomenclature. An idea is to default to a different dichotomy: 

- Originating vs Receiving
- Expensive vs Cheaper

An edge case that I see from this dichotomy is a Originating chain that is Cheaper, so most of the validations is done on the Cheaper chain

File: docs/bridges/tokenbridge/amb-bridge.md [hunk 9]
```diff
@@ -57,16 +52,104 @@ References:
...
+4. `executeAffirmation()` is called on the Home Bridge contract by a validator once enough signatures are collected. 
+5. Home bridge contract decodes the message and calls `foo()` on the target contract. 
+#### Home Network -> Foreign Network
+1. User calls `foo()` on an originator contract
+2. Originator contract calls `requireToPassMessage()` on Home Bridge contract, and encodes `foo()`, target address, and includes some tokens for gas.
+3. Signatures are collected from validators, and once enough are collected `requireToConfirmMessage()` is called
+4. Message is relayed to the Foreign Bridge contract, and `executeSignatures()` is called
+5. Foreign bridge contract decodes the message and calls `foo()` on target contract
+
+AMB Bridge proxy contact on Ethereum can be viewed [here](https://etherscan.io/address/0x4C36d2919e407f0Cc2Ee3c993ccF8ac26d9CE64e)
```
Comment (dan-gnosis) [inline docs/bridges/tokenbridge/amb-bridge.md:103]: Avoid using "here" for hyperlinks - it's not SEO-friendly. 

I would advise to just link [AMB Bridge Proxy Contract on Ethereum](https://etherscan.io/address/0x4C36d2919e407f0Cc2Ee3c993ccF8ac26d9CE64e) directly. 

Also, when linking to Etherscan or Gnosisscan, default to the #code or #writeProxyContract page, which shows the underlying methods. 

image

File: docs/bridges/tokenbridge/amb-bridge.md [hunk 10]
```diff
@@ -57,16 +52,104 @@ References:
...
+5. Home bridge contract decodes the message and calls `foo()` on the target contract. 
+#### Home Network -> Foreign Network
+1. User calls `foo()` on an originator contract
+2. Originator contract calls `requireToPassMessage()` on Home Bridge contract, and encodes `foo()`, target address, and includes some tokens for gas.
+3. Signatures are collected from validators, and once enough are collected `requireToConfirmMessage()` is called
+4. Message is relayed to the Foreign Bridge contract, and `executeSignatures()` is called
+5. Foreign bridge contract decodes the message and calls `foo()` on target contract
+
+AMB Bridge proxy contact on Ethereum can be viewed [here](https://etherscan.io/address/0x4C36d2919e407f0Cc2Ee3c993ccF8ac26d9CE64e)  
+AMB Bridge Proxy Contract and Gnosis can be viewed [here](https://blockscout.com/xdai/mainnet/address/0x75Df5AF045d91108662D8080fD1FEFAd6aA0bb59)
```
Comment (dan-gnosis) [inline docs/bridges/tokenbridge/amb-bridge.md:104]: See above

File: docs/bridges/tokenbridge/amb-bridge.md [hunk 11]
```diff
@@ -57,16 +52,104 @@ References:
...
+| Authorization | Check that `msg.sender` is the address of the bridge contract                                                                                                                                                                                                                                                                  |
+| Replay Attack | `transactionHash()` allows for checking of a hash of the transaction that invoked the `requireToPassMessage()` call. The invoking contract (in some cases, the mediator contract) is responsible for providing a *unique sequence* (can be a nonce) as part of the `_data` param in the `requireToPassMessage()` function call |
+
+### Mediator Contracts
+ 
+ A mediator contract is needed if there is an approval flow, such as when transferring an NFT or ERC-20. For a more in-depth explanation, see the [Omnibridge page](omnibridge).
```
Comment (dan-gnosis) [inline docs/bridges/tokenbridge/amb-bridge.md:114]: I'm not sure if the Omnibridge page explains the Mediator Contracts at all

File: docs/bridges/tokenbridge/amb-bridge.md [hunk 12]
```diff
@@ -57,16 +52,104 @@ References:
...
+|------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| System Contracts | AMB Implementation Contracts (Home Bridge and Foreign Bridge), Governance Multisigs, gas limit helpers, failed call management helpers (for when gas estimate was insufficient), and fee management helpers to collect fees |
+| Oracles | Containerized microservices that listen for on-chain events and send confirmations to relay messages. [More on them here](https://github.com/omni/tokenbridge/blob/master/oracle/README.md).
+| DevOps | [Bridge monitoring](https://github.com/omni/tokenbridge/blob/master/monitor/README.md), [ALM](https://github.com/omni/tokenbridge/tree/master/alm), docker compose, ansible playbooks |
+| dApp Contracts | extensions (pair mediator contracts on both sides of the AMB), such as the Omnibridge |
+
+#### Bridge Modes
```
Comment (dan-gnosis) [inline docs/bridges/tokenbridge/amb-bridge.md:125]: I recommend we turn this into a Heading 3, and describe it differently: 

### "Uses of AMB"

- ERC-to-ERC Bridges: ...
- ERC-to-Native Bridges: ...
- Message Passing: ...

**Review by dan-gnosis — CHANGES_REQUESTED** [review]
Good work. Some changes but headed in the right direction.

File: docs/bridges/tokenbridge/amb-bridge.md (hunk 5 above)
Comment (CJBangert) [reply docs/bridges/tokenbridge/amb-bridge.md:88]: Yep, my understanding is the mediator contract would call that function on the bridge contract

File: docs/bridges/tokenbridge/amb-bridge.md (hunk 8 above)
Comment (CJBangert) [reply docs/bridges/tokenbridge/amb-bridge.md:92]: Yes, this is the same terminology as above. I've just stuck to using the terminology used in the code up through now but I agree it is confusing

File: docs/bridges/tokenbridge/amb-bridge.md (hunk 11 above)
Comment (CJBangert) [reply docs/bridges/tokenbridge/amb-bridge.md:114]: Working across 2 branches, going to go back through and add a sub-heading to link to

File: docs/bridges/tokenbridge/amb-bridge.md (hunk 7 above)
Comment (CJBangert) [reply docs/bridges/tokenbridge/amb-bridge.md]: I changed it to "originating contract" and added a definition under terminology. This is just an arbitrary contract that the user interacts with where the message originates.

**Comment by CJBangert** [comment]
link will be broken until omnibridge branch is merged

**Review by dan-gnosis — APPROVED** [review]

**Pull request merged**